File: server/services/auth.service.js
/**
 * Authentication Service
 * Handles user authentication, registration, and token management
 */

const { admin, auth, db } = require('../config/firebase-admin');
const config = require('../config/config');
const jwt = require('jsonwebtoken');

/**
 * Create a new user account
 * @param {string} email - User email
 * @param {string} password - User password
 * @param {string} displayName - User display name
 * @returns {Promise<Object>} - Created user data
 */
const createUser = async (email, password, displayName) => {
  try {
    // Create the user in Firebase Authentication
    const userRecord = await auth.createUser({
      email,
      password,
      displayName,
      emailVerified: false
    });

    // Generate verification email link
    const emailVerificationLink = await auth.generateEmailVerificationLink(email);

    // Create user document in Firestore
    const userData = {
      email,
      displayName,
      photoURL: null,
      accountTier: 'free',
      status: 'active',
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
      updatedAt: admin.firestore.FieldValue.serverTimestamp(),
      lastLoginAt: admin.firestore.FieldValue.serverTimestamp(),
      moduleCount: 0,
      noteCount: 0,
      storageUsed: 0
    };

    await db.collection('users').doc(userRecord.uid).set(userData);

    return {
      uid: userRecord.uid,
      emailVerificationLink,
      ...userData
    };
  } catch (error) {
    throw error;
  }
};

/**
 * Login user and generate auth token
 * @param {string} email - User email
 * @param {string} password - User password
 * @returns {Promise<Object>} - User data and token
 */
const loginUser = async (email, password) => {
  try {
    // Verify user credentials with Firebase Auth
    const userCredential = await auth.getUserByEmail(email);
    
    // Check if user exists and is active
    const userDoc = await db.collection('users').doc(userCredential.uid).get();
    
    if (!userDoc.exists) {
      throw new Error('User record not found');
    }
    
    const userData = userDoc.data();
    
    if (userData.status === 'deleted') {
      throw new Error('User account has been deleted');
    }
    
    // Update last login timestamp
    await db.collection('users').doc(userCredential.uid).update({
      lastLoginAt: admin.firestore.FieldValue.serverTimestamp()
    });
    
    // Generate JWT token
    const token = jwt.sign(
      { 
        uid: userCredential.uid,
        email: userCredential.email,
        tier: userData.accountTier 
      }, 
      config.jwt.secret, 
      { expiresIn: config.jwt.expiresIn }
    );
    
    return {
      user: {
        uid: userCredential.uid,
        email: userCredential.email,
        displayName: userData.displayName,
        photoURL: userData.photoURL,
        accountTier: userData.accountTier
      },
      token
    };
  } catch (error) {
    throw error;
  }
};

/**
 * Send password reset email
 * @param {string} email - User email
 * @returns {Promise<string>} - Password reset link
 */
const sendPasswordResetEmail = async (email) => {
  try {
    const resetLink = await auth.generatePasswordResetLink(email);
    return resetLink;
  } catch (error) {
    throw error;
  }
};

/**
 * Verify JWT token
 * @param {string} token - JWT token
 * @returns {Object} - Decoded token payload
 */
const verifyToken = (token) => {
  try {
    const decoded = jwt.verify(token, config.jwt.secret);
    return decoded;
  } catch (error) {
    throw error;
  }
};

/**
 * Get user profile data
 * @param {string} uid - User ID
 * @returns {Promise<Object>} - User profile data
 */
const getUserProfile = async (uid) => {
  try {
    const userDoc = await db.collection('users').doc(uid).get();
    
    if (!userDoc.exists) {
      throw new Error('User not found');
    }
    
    const userData = userDoc.data();
    
    if (userData.status === 'deleted') {
      throw new Error('User account has been deleted');
    }
    
    return {
      uid: userDoc.id,
      ...userData
    };
  } catch (error) {
    throw error;
  }
};

/**
 * Soft delete a user account
 * @param {string} uid - User ID
 * @returns {Promise<void>}
 */
const deleteUser = async (uid) => {
  try {
    // Soft delete by updating user status
    await db.collection('users').doc(uid).update({
      status: 'deleted',
      deletedAt: admin.firestore.FieldValue.serverTimestamp()
    });
  } catch (error) {
    throw error;
  }
};

module.exports = {
  createUser,
  loginUser,
  sendPasswordResetEmail,
  verifyToken,
  getUserProfile,
  deleteUser
};

File: notetaker_frontend/lib/services/auth_service.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

class AuthService {
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  
  // Get current user
  User? get currentUser => _auth.currentUser;
  
  // Auth state changes stream
  Stream<User?> get authStateChanges => _auth.authStateChanges();
  
  // Sign in with email and password
  Future<User?> signInWithEmailAndPassword(String email, String password) async {
    try {
      final userCredential = await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
      
      // Update last login timestamp
      if (userCredential.user != null) {
        await _firestore.collection('users').doc(userCredential.user!.uid).update({
          'lastLoginAt': FieldValue.serverTimestamp(),
        });
      }
      
      return userCredential.user;
    } catch (e) {
      print('Login error: $e');
      rethrow;
    }
  }
  
  // Create a new user
  Future<User?> createUserWithEmailAndPassword(
    String email, 
    String password, 
    String displayName
  ) async {
    try {
      final userCredential = await _auth.createUserWithEmailAndPassword(
        email: email,
        password: password,
      );
      
      // Update display name
      if (userCredential.user != null) {
        await userCredential.user!.updateDisplayName(displayName);
        
        // Initialize user document in Firestore
        await _firestore.collection('users').doc(userCredential.user!.uid).set({
          'email': email,
          'displayName': displayName,
          'photoURL': null,
          'accountTier': 'free',
          'moduleCount': 0,
          'noteCount': 0,
          'storageUsed': 0,
          'createdAt': FieldValue.serverTimestamp(),
          'updatedAt': FieldValue.serverTimestamp(),
          'lastLoginAt': FieldValue.serverTimestamp(),
        });
      }
      
      return userCredential.user;
    } catch (e) {
      print('Registration error: $e');
      rethrow;
    }
  }
  
  // Sign out
  Future<void> signOut() async {
    await _auth.signOut();
  }
  
  // Reset password
  Future<void> resetPassword(String email) async {
    await _auth.sendPasswordResetEmail(email: email);
  }
  
  // Get user data from Firestore
  Future<Map<String, dynamic>> getUserData(String userId) async {
    try {
      final documentSnapshot = await _firestore.collection('users').doc(userId).get();
      
      if (documentSnapshot.exists) {
        return documentSnapshot.data() ?? {};
      } else {
        return {};
      }
    } catch (e) {
      print('Error getting user data: $e');
      return {};
    }
  }
  
  // Update user profile
  Future<void> updateUserProfile(String userId, Map<String, dynamic> data) async {
    await _firestore.collection('users').doc(userId).update({
      ...data,
      'updatedAt': FieldValue.serverTimestamp(),
    });
  }
}